content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+Of course the external API call just by itself adds to the request latency,
+which is something we might want to avoid.
+
+It's also worth examining the applications garbage collection behavior.
+In order to so, you can run the performance test method again,
+either enabling all the GC-related JFR event types, or by enabling a pre-existing JFR configuration
+(the JDK comes with two built-in JFR configurations, and you can also create and export them via JMC):
+
+[source, java]
+----
+@Test
+@EnableConfiguration("profile")
+public void retrieveTodo() throws Exception {
+  // ...
+}
+----
+
+Open the recording in JMC, and you'll see there's a substantial amount of GC activity happening:
+
+image::continuous_perf_testing_gc_regression.png[Garbage collections after the performance regression]
+
+The difference to the GC behavior before this code change is striking:
+
+image::continuous_perf_testing_gc_original.png[Garbage collections before the performance regression]
+
+Pause times are worse, directly impacting the application's latency, and the largely increased GC volume means the environment will be able to serve less concurrent requests when reaching its capacity limits,
+meaning you'd have to provision another machine earlier on as your load increases.
+
+[NOTE]
+.Memory Leak in the JFR Event Streaming API
+====
+The astute reader may have noticed that there is a memory leak before _and_ after the code change, as indicated by the ever increased heap size post GC.
+After some exploration https://mail.openjdk.java.net/pipermail/hotspot-jfr-dev/2020-December/001972.html[it turned out] that this is a bug in the JFR event streaming API which holds on to a large number of `RecordedEvent` instances internally.
+https://twitter.com/ErikGahlin[Erik Gahlin] from the OpenJDK team logged https://bugs.openjdk.java.net/browse/JDK-8257906[JDK-8257906] for tracking and hopefully fixing this in JDK 16.
+====
+
+Now such drastic increase of allocation and thus potential impact on performance should hopefully be an exception rather than a regular situation.
+But the example shows how continuous performance unit tests on impacting metrics like memory allocation via Flight Recorder and JfrUnit can help to identify performance issues in an automated and reliable way,
+preventing such regression to sneak into production.
+Being able to identify this kind of issue by running tests locally on a developer laptop or a CI server,
+can be a huge time-saver and productivity boost.
+
+== Identifying Increased I/O With the Database
+
+Once you've started to look at performance tests through the lense of JfrUnit,
+more and more possibilities pop up.
+Asserting a maximum number of garbage collections? Not a problem.
+Avoiding an unexpected amount of file system IO? The `jdk.FileRead` and `jdk.FileWrite` events are our friend.
+Examining and asserting the I/O done with the database? Easily doable.
```

Comment:
I think I'd like also:

1. implement your own custom JFR events
2. use this to test your custom events

:-)

Reply:
Done.